libp2p-xtra/src/substream.rs

```diff
@@ -0,0 +1,158 @@
+pub struct Substream {
+    #[pin]
+    inner: Negotiated<yamux::Stream>,
+
+    /// The prometheus timer tracking the duration of the substream.
+    ///
+    /// This timer is started upon construction and automatically stops once it is dropped. Thus,
+    /// we only need to keep it around but not interact with it after construction.
+    _timer: HistogramTimer,
+
+    /// The prometheus counter for the number of bytes read.
+    read_counter: IntCounter,
+
+    /// The prometheus counter for the number of bytes written.
+    written_counter: IntCounter,
+}
+
+impl Substream {
+    pub(crate) fn new(inner: Negotiated<yamux::Stream>, protocol: &'static str) -> Self {
+        let labels = HashMap::from([(PROTOCOL_LABEL, protocol)]);
+
+        Self {
+            inner,
+            _timer: SUBSTREAM_DURATION_HISTOGRAM.with(&labels).start_timer(),
+            read_counter: SUBSTREAM_BYTES_READ_COUNTER.with(&labels),
+            written_counter: SUBSTREAM_BYTES_WRITTEN_COUNTER.with(&labels),
+        }
+    }
+}
+
+impl AsyncRead for Substream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::io::Result<usize>> {
+        let this = self.project();
+
+        let bytes_read = ready!(this.inner.poll_read(cx, buf)?);
+
+        this.read_counter.inc_by(bytes_read as u64);
+
+        Poll::Ready(Ok(bytes_read))
+    }
+
+    fn poll_read_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &mut [IoSliceMut<'_>],
+    ) -> Poll<std::io::Result<usize>> {
+        let this = self.project();
+
+        let bytes_read = ready!(this.inner.poll_read_vectored(cx, bufs)?);
+        this.read_counter.inc_by(bytes_read as u64);
+
+        Poll::Ready(Ok(bytes_read))
+    }
+}
+
+impl AsyncWrite for Substream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        let this = self.project();
+
+        let bytes_written = ready!(this.inner.poll_write(cx, buf)?);
+        this.written_counter.inc_by(bytes_written as u64);
+
+        Poll::Ready(Ok(bytes_written))
+    }
+
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<std::io::Result<usize>> {
+        let this = self.project();
+
+        let bytes_written = ready!(this.inner.poll_write_vectored(cx, bufs)?);
+        this.written_counter.inc_by(bytes_written as u64);
+
+        Poll::Ready(Ok(bytes_written))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        let this = self.project();
+
+        this.inner.poll_flush(cx)
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        let this = self.project();
+
+        this.inner.poll_close(cx)
+    }
+}
+
+const PROTOCOL_LABEL: &str = "protocol";
```

Comment:
❓ Does this mean that every protocol will have the same label? Shouldn't this be configurable?

Reply:
This is the label "key"! The labels are a `HashMap`. One label can have as many values (the actual protocol names) as you want but you need to specify the labels upfront!